io/src/main/scala/fs2/io/io.scala

```diff
@@ -146,9 +156,24 @@ package object io {
 
     Stream.resource(mkOutput).flatMap {
       case (os, is) =>
-        val write = f(os) *> Sync[F].delay(os.close())
-        val read = readInputStream(is.pure[F], chunkSize, blocker, closeAfterUse = false)
-        read.concurrently(Stream.eval(write))
+        Stream.eval(Ref.of[F, Option[Throwable]](None)).flatMap { ref =>
+          // We need to close the output stream regardless of how `f` finishes
+          // to ensure an outstanding blocking read on the input stream completes.
+          // In such a case, there's a race between completion of the read
+          // stream and finalization of the write stream, so we capture the error
+          // that occurs when writing and rethrow it.
+          val write = f(os).guaranteeCase(
+            ec =>
+              blocker.delay(os.close()) *> ref.set(ec match {
+                case ExitCase.Error(t) => Some(t); case _ => None
+              })
+          )
+          val read = readInputStream(is.pure[F], chunkSize, blocker, closeAfterUse = false)
+          read.concurrently(Stream.eval(write)) ++ Stream.eval(ref.get).flatMap {
+            case None    => Stream.empty
+            case Some(t) => Stream.raiseError[F](t)
+          }
```

Comment:
This is messy... How about something like this:
```scala
val mkIn = Resource.fromAutoCloseable(Sync[F].delay { new PipedInputStream })
def mkOut(is: PipedInputStream) = Resource.fromAutoCloseable(Sync[F].delay { new PipedOutputStream(is) })

Stream.resource(mkIn).flatMap { is =>
  Deferred.of[F, Unit].flatMap { connected =>
    val write = Stream.eval_(mkOut(is).evalTap(connected.complete(())).use(f))
    // We can't read from the pipe until we are sure it will definitely get closed.
    val read = Stream.eval_(connected.get) ++ readInputStream(is.pure[F], chunkSize, blocker, closeAfterUse = false)
    read.concurrently(write)
  }
}
```
This still has annoying synchronization (I don't see how we can get rid of it), but no explicit closing or error handling. It's relatively clear that reading from the pipe doesn't prevent the `PipedOutputStream` from closing. I can't see a deadlock, either - although it's already night here, so I'm not at my best.

Reply:
Yeah, I like yours better. Want to open a PR and I'll close this one?

Reply:
I'd rather it was incorporated into yours in any way you see fit, if that's all right with you.

Reply:
:+1: Just pushed an update.

Reply:
I had to revert it as the new implementation still has a race at shutdown time:

```
- fails when `f` fails *** FAILED *** (90 milliseconds)
[info]   GeneratorDrivenPropertyCheckFailedException was thrown during property evaluation. (IoSpec.scala:54)
[info]     Falsified after 10 successful property evaluations.
[info]     Message: Right(Vector()) was not equal to Left(java.lang.Exception: boom)
[info]     Location: (IoSpec.scala:58)
[info]     Occurred when passed generated values (
[info]       10
[info]     )
[info]     Init Seed: 1571079680891
...
Cause: org.scalatest.exceptions.TestFailedException: Right(Vector()) was not equal to Left(java.lang.Exception: boom)
```